stdlib/public/core/StringCreate.swift

```diff
@@ -298,4 +298,53 @@ extension String {
       String._uncheckedFromUTF8($0)
     }
   }
+
+  @usableFromInline
+  @available(SwiftStdlib 5.10, *)
+  internal static func _validate<Encoding: Unicode.Encoding>(
+    _ input: UnsafeBufferPointer<Encoding.CodeUnit>,
+    as encoding: Encoding.Type
+  ) -> String? {
+  fast: // fast-path
+    if encoding.CodeUnit.self == UInt8.self {
+      let bytes = _identityCast(input, to: UnsafeBufferPointer<UInt8>.self)
+      let isASCII: Bool
+      if encoding.self == UTF8.self {
+        guard case .success(let info) = validateUTF8(bytes) else { return nil }
+        isASCII = info.isASCII
+      } else if encoding.self == Unicode.ASCII.self {
+        guard _allASCII(bytes) else { return nil }
+        isASCII = true
+      } else {
+        break fast
+      }
+      return String._uncheckedFromUTF8(bytes, asciiPreScanResult: isASCII)
+    }
+
+    // slow-path
+    // this multiplier is a worst-case estimate
+    let multiplier = if encoding.self == UTF16.self { 3 } else { 4 }
+    return withUnsafeTemporaryAllocation(
+      of: UInt8.self, capacity: input.count * multiplier
+    ) {
```

Comment:
This is a generic function - transcoding will ultimately be performed by a user-defined type, and if that type were to produce _more_ than 4 bytes of UTF8 for its custom code-units, this would over-write the buffer and corrupt the stack.

Reply:
Toy example, showing that the `Unicode.Encoding` API allows this. This encoding is UTF8, except that every time the _real_ UTF8 would parse a single scalar from its code-units, this encoding parses 5 repetitions of that scalar:

https://gist.github.com/karwa/ece9cdcf8c66613fdeea85bcd8b2cea8

Reply:
Good point. We should bounds-check here. We should write directly into a string buffer, reallocating when appropriate.